File: src/IECore/InterpolatedCache.cpp
//////////////////////////////////////////////////////////////////////////
//
//  Copyright (c) 2007-2009, Image Engine Design Inc. All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are
//  met:
//
//     * Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//
//     * Redistributions in binary form must reproduce the above copyright
//       notice, this list of conditions and the following disclaimer in the
//       documentation and/or other materials provided with the distribution.
//
//     * Neither the name of Image Engine Design nor the names of any
//       other contributors to this software may be used to endorse or
//       promote products derived from this software without specific prior
//       written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
//////////////////////////////////////////////////////////////////////////

#include <cassert>

#include "boost/format.hpp"

#include "IECore/MessageHandler.h"
#include "IECore/OversamplesCalculator.h"
#include "IECore/ObjectInterpolator.h"
#include "IECore/InterpolatedCache.h"
#include "IECore/CompoundObject.h"
#include "IECore/FileSequence.h"
#include "IECore/EmptyFrameList.h"

using namespace IECore;
using namespace boost;

InterpolatedCache::InterpolatedCache( const std::string &pathTemplate, float frame, Interpolation interpolation, const OversamplesCalculator &o )
		: m_fileSequence( 0 ), m_oversamplesCalculator( o )
{
	setPathTemplate( pathTemplate );
	setFrame( frame );
	setInterpolation( interpolation );

	m_parametersChanged = true;
}

InterpolatedCache::~InterpolatedCache()
{
	closeCacheFiles();
}

void InterpolatedCache::setPathTemplate( const std::string &pathTemplate )
{
	if ( !m_fileSequence || getPathTemplate() != pathTemplate )
	{
		m_fileSequence = new FileSequence( pathTemplate, new EmptyFrameList() );
		m_parametersChanged = true;
		closeCacheFiles();
	}
}

const std::string &InterpolatedCache::getPathTemplate() const
{
	assert( m_fileSequence );
	return m_fileSequence->getFileName();
}

void InterpolatedCache::setFrame( float frame )
{
	if ( frame != m_frame )
	{
		m_frame = frame;
		m_parametersChanged = true;
	}
}

float InterpolatedCache::getFrame() const
{
	return m_frame;
}

void InterpolatedCache::setInterpolation( InterpolatedCache::Interpolation interpolation )
{
	if ( interpolation != m_interpolation )
	{
		m_interpolation = interpolation;
		m_parametersChanged = true;
	}
}

InterpolatedCache::Interpolation InterpolatedCache::getInterpolation() const
{
	return m_interpolation;
}

void InterpolatedCache::setOversamplesCalculator( const OversamplesCalculator &oc )
{
	m_oversamplesCalculator = oc;
}

OversamplesCalculator InterpolatedCache::getOversamplesCalculator() const
{
	return m_oversamplesCalculator;
}


ObjectPtr InterpolatedCache::read( const ObjectHandle &obj, const AttributeHandle &attr ) const
{
	updateCacheFiles();

	ObjectPtr data;

	if ( m_useInterpolation )
	{
		std::vector< ObjectPtr > pts;
		for ( unsigned int c = 0; c < m_caches.size(); c++ )
		{
			data = m_caches[c]->read( obj, attr );
			assert( data );
			pts.push_back( data );
		}
		assert( pts.size() == m_caches.size() );

		switch ( m_interpolation )
		{
		case Linear:
		{
			assert( pts.size() == 2 );
			data = linearObjectInterpolation( pts[0], pts[1], m_x );

			break;
		}
		case Cubic:
		{
			assert( pts.size() == 4 );
			data = cubicObjectInterpolation( pts[0], pts[1], pts[2], pts[3], m_x );
			break;
		}
		default:
			assert( false );
		}

		if ( !data )
		{
			data = pts[ m_curFrameIndex ];
		}
	}
	else
	{
		data = m_caches[ m_curFrameIndex ]->read( obj, attr );
	}

	assert( data );
	return data;
}

CompoundObjectPtr InterpolatedCache::read( const ObjectHandle &obj ) const
{
	CompoundObjectPtr dict = new CompoundObject();
	std::vector<AttributeHandle> attrs;
	attributes( obj, attrs );

	for ( std::vector<AttributeHandle>::const_iterator it = attrs.begin(); it != attrs.end(); ++it )
	{
		ObjectPtr data = read( obj, *it );

		dict->members()[ *it ] = data;
	}

	return dict;
}

ObjectPtr InterpolatedCache::readHeader( const HeaderHandle &hdr ) const
{
	updateCacheFiles();

	ObjectPtr data;

	if ( m_useInterpolation )
	{
		std::vector< ObjectPtr > pts;
		for ( unsigned int c = 0; c < m_caches.size(); c++ )
		{
			data = m_caches[c]->readHeader( hdr );
			assert( data );
			pts.push_back( data );
		}
		assert( pts.size() == m_caches.size() );

		switch ( m_interpolation )
		{
		case Linear:
		{
			assert( pts.size() == 2 );
			data = linearObjectInterpolation( pts[0], pts[1], m_x );
			break;
		}
		case Cubic:
		{
			assert( pts.size() == 4 );
			data = cubicObjectInterpolation( pts[0], pts[1], pts[2], pts[3], m_x );
			break;
		}
		default:
			assert( false );
		}

		if ( !data )
		{
			data = pts[ m_curFrameIndex ];
		}
	}
	else
	{
		data = m_caches[ m_curFrameIndex ]->readHeader( hdr );
	}

	assert( data );
	return data;
}

CompoundObjectPtr InterpolatedCache::readHeader( ) const
{
	updateCacheFiles();

	CompoundObjectPtr dict = new CompoundObject();
	std::vector<HeaderHandle> hds;
	headers( hds );

	for ( std::vector<HeaderHandle>::const_iterator it = hds.begin(); it != hds.end(); ++it )
	{
		ObjectPtr data = readHeader( *it );

		dict->members()[ *it ] = data;

	}
	return dict;
}

void InterpolatedCache::objects( std::vector<ObjectHandle> &objs ) const
{
	updateCacheFiles();
	assert( m_caches.size() > m_curFrameIndex );
	m_caches[ m_curFrameIndex ]->objects( objs );
}

void InterpolatedCache::headers( std::vector<HeaderHandle> &hds ) const
{
	updateCacheFiles();
	assert( m_caches.size() > m_curFrameIndex );
	m_caches[ m_curFrameIndex ]->headers( hds );
}

void InterpolatedCache::attributes( const ObjectHandle &obj, std::vector<AttributeHandle> &attrs ) const
{
	updateCacheFiles();
	assert( m_caches.size() > m_curFrameIndex );
	m_caches[ m_curFrameIndex ]->attributes( obj, attrs );
}

void InterpolatedCache::attributes( const ObjectHandle &obj, const std::string regex, std::vector<AttributeHandle> &attrs ) const
{
	updateCacheFiles();
	assert( m_caches.size() > m_curFrameIndex );
	m_caches[ m_curFrameIndex ]->attributes( obj, regex, attrs );
}

bool InterpolatedCache::contains( const ObjectHandle &obj ) const
{
	updateCacheFiles();
	assert( m_caches.size() > m_curFrameIndex );
	return m_caches[ m_curFrameIndex ]->contains( obj );
}

bool InterpolatedCache::contains( const ObjectHandle &obj, const AttributeHandle &attr ) const
{
	updateCacheFiles();
	assert( m_caches.size() > m_curFrameIndex );
	return m_caches[ m_curFrameIndex ]->contains( obj, attr );
}

void InterpolatedCache::updateCacheFiles() const
{
	if ( !m_parametersChanged )
	{
		return;
	}

	int lowTick, highTick;
	float x = m_oversamplesCalculator.tickInterval( m_frame, lowTick, highTick );

	float step = ( float )m_oversamplesCalculator.getTicksPerSecond() / m_oversamplesCalculator.getSamplesPerFrame() / m_oversamplesCalculator.getFrameRate() ;

	int start = 0;
	int end = 0;

	if ( x == 0. || m_interpolation == None )
	{
		m_useInterpolation = false;
		start = 0;
		end = 0;
		m_curFrameIndex = 0;
	}
	else
	{
		m_useInterpolation = true;
		switch ( m_interpolation )
		{
		case Linear:
			// Need one file ahead
			m_curFrameIndex = 0;
			start = 0;
			end = 1;
			break;
		case Cubic:
			// Need one file behind, 2 files ahead
			m_curFrameIndex = 1;
			start = -1;
			end = 2;
			break;
		default:
			assert( false );
		}
	}

	CacheVector caches;
	std::vector< std::string > cacheFiles;
	AttributeCachePtr cache;
	// Open all the cache files required to perform interpolation
	for ( int fileNum = start; fileNum <= end; fileNum ++ )
	{
		int tick = m_oversamplesCalculator.nearestTick(( int )( lowTick + fileNum * step ) );
		std::string fileName = m_fileSequence->fileNameForFrame( tick );

		cache = 0;
		for ( unsigned i = 0; i < m_caches.size(); i++ )
		{
			if ( fileName == m_cacheFiles[i] )
			{
				cache = m_caches[i];
				break;
			}
		}
		if ( !cache )
		{
			cache = new AttributeCache( fileName, IndexedIO::Read );
		}

		assert( cache );
		cacheFiles.push_back( fileName );
		caches.push_back( cache );
	}

	// save info in the object.
	m_x = x;
	m_caches = caches;
	m_cacheFiles = cacheFiles;
	assert( m_caches.size() == m_cacheFiles.size() );
	assert( m_caches.size() > m_curFrameIndex );
	m_parametersChanged = false;
}

void InterpolatedCache::closeCacheFiles() const
{
	m_parametersChanged = true;
	m_caches.clear();
	m_cacheFiles.clear();
}

File: contrib/removedInCortex9/attributeCache/test/IECore/InterpolatedCache.py
##########################################################################
#
#  Copyright (c) 2007-2013, Image Engine Design Inc. All rights reserved.
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are
#  met:
#
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#
#     * Redistributions in binary form must reproduce the above copyright
#       notice, this list of conditions and the following disclaimer in the
#       documentation and/or other materials provided with the distribution.
#
#     * Neither the name of Image Engine Design nor the names of any
#       other contributors to this software may be used to endorse or
#       promote products derived from this software without specific prior
#       written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
##########################################################################


"""Unit test for InterpolatedCache binding"""
import os, glob, random
import unittest
import math
import threading

from IECore import *


class TestInterpolatedCache(unittest.TestCase):

	pathTemplate = "./test/cache_####.fio"

	def mark( self ) :

		self.__createCacheFile( 250, 0 )
		self.__createCacheFile( 333, 1 )
		self.__createCacheFile( 416, 2 )
		self.__createCacheFile( 500, 3 )
		self.__createCacheFile( 583, 4 )
		self.__createCacheFile( 666, 5 )
		self.__createCacheFile( 750, 6 )

	def __createV3f( self, value ):
		dataWritten = V3fVectorData()
		for i in range(0, 3):
			dataWritten.append( V3f( value, value, value ) )
		return dataWritten

	def __createCacheFile( self, frame, value ):

		fs = FileSequence( self.pathTemplate, EmptyFrameList() )

		cache = AttributeCache( fs.fileNameForFrame( frame ), IndexedIO.OpenMode.Write)
	
		dataWritten = self.__createV3f( value )
		cache.write( "obj1", "v3fVec", dataWritten)
		cache.write( "obj2", "i", IntData( value ) )
		cache.write( "obj2", "d", DoubleData( value ) )
		cache.writeHeader( "testCache", self.__headerCompound( frame ) )

	def __headerCompound( self, frame ):
		return CompoundObject(
			{
				"a": IntData( 1 ),
				"b": StringData( "HEllo!" ),
				"frame": FloatData( frame ),
			}
		)

	def __time( self, frame ):
		return int( frame * 6000. / 24. )

	def __createCache( self ):
		"""Create some cache files do play with..."""

		self.__createCacheFile( self.__time( 0 ), 0 )
		self.__createCacheFile( self.__time( 1 ), 1 )
		self.__createCacheFile( self.__time( 2 ), 2 )
		self.__createCacheFile( self.__time( 3 ), 4 )
		self.__createCacheFile( self.__time( 4 ), 8 )
		self.__createCacheFile( self.__time( 5 ), 16 )

	def testConstructors(self):
		"""Test InterpolatedCache constructors"""
		cache = InterpolatedCache(self.pathTemplate, interpolation = InterpolatedCache.Interpolation.None )
		self.assertEqual( cache.getInterpolation(), InterpolatedCache.Interpolation.None )
		self.assertEqual( cache.getPathTemplate(), self.pathTemplate )

		os = OversamplesCalculator( frameRate = 20, samplesPerFrame = 1 )
		cache = InterpolatedCache(self.pathTemplate, interpolation = InterpolatedCache.Interpolation.Linear, oversamplesCalculator = os )
		self.assertEqual( cache.getInterpolation(), InterpolatedCache.Interpolation.Linear )

		os = OversamplesCalculator( frameRate = 24, samplesPerFrame = 1 )
		cache = InterpolatedCache(self.pathTemplate, interpolation = InterpolatedCache.Interpolation.Cubic, oversamplesCalculator = os )

	def testHeaders( self ):
		"""Test InterpolatedCache headers"""

		self.__createCache()

		cache = InterpolatedCache(self.pathTemplate, interpolation = InterpolatedCache.Interpolation.Linear )
		self.assert_( "testCache" in cache.headers( 1.5 ) )
		h = cache.readHeader( 1.5, "testCache" )
		self.assertEqual( h, self.__headerCompound( ( self.__time( 1 ) + self.__time( 2 ) ) / 2 ) )

	def testAttributes(self):
		"""Test InterpolatedCache attributes"""

		self.__createCache()

		cache = InterpolatedCache(self.pathTemplate, interpolation = InterpolatedCache.Interpolation.Linear )
		self.assertEqual( len( cache.attributes( 1.2, "obj2" )), 2 )
		self.assertEqual( len( cache.attributes( 1.2, "obj2", "i.*" )), 1 )
		self.assertEqual( len( cache.attributes( 1.2, "obj1" )), 1 )

	def testContains(self):
		"""Test InterpolatedCache contains"""

		self.__createCache()

		cache = InterpolatedCache(self.pathTemplate, interpolation = InterpolatedCache.Interpolation.Linear )
		self.assert_( cache.contains( 1.2, "obj1" ) )
		self.assert_( cache.contains( 1.2, "obj2", "i" ) )
		self.failIf( cache.contains( 1.2, "obj3" ) )
		self.failIf( cache.contains( 1.2, "obj2", "not_in_cache" ) )

	def testReading(self):
		"""Test InterpolatedCache read"""

		self.__createCache()

		cache = InterpolatedCache( self.pathTemplate, interpolation = InterpolatedCache.Interpolation.Linear )
		self.assertEqual( cache.read( 1.5, "obj2", "i" ), IntData( 1 ) )
		self.assertEqual( cache.read( 1.5, "obj2", "d" ), DoubleData( 1.5 ) )
		self.assertEqual( cache.read( 1.5, "obj1" ), CompoundObject( { "v3fVec": self.__createV3f( 1.5 ) } ) )

	def testOversampledReading( self ) :

		self.mark()

		os = OversamplesCalculator( frameRate = 24, samplesPerFrame = 3 )
		cache = InterpolatedCache(self.pathTemplate, interpolation = InterpolatedCache.Interpolation.Linear, oversamplesCalculator = os )
		self.assertAlmostEqual( cache.read( 2, "obj2", "d" ).value, 3, 4 )


	def testChangingParameters( self ):
		"""Test InterpolatedCache reading and changing frame, interpolation, ... """

		self.__createCache()

		cache = InterpolatedCache(self.pathTemplate, interpolation = InterpolatedCache.Interpolation.None )
		self.assertEqual( cache.read( 0, "obj2", "d" ), DoubleData( 0 ) )
		self.assertEqual( cache.read( 0.8, "obj2", "d" ), DoubleData( 0 ) )
		self.assertEqual( cache.read( 1, "obj2", "d" ), DoubleData( 1 ) )
		cache.setInterpolation( InterpolatedCache.Interpolation.Linear )
		self.assertEqual( cache.read( 0.5, "obj2", "d" ), DoubleData( 0.5 ) )
		self.assertAlmostEqual( cache.read( 2.25, "obj2", "d" ).value, 2.5, 3 )
		cache.setInterpolation( InterpolatedCache.Interpolation.Cubic )
		self.assertAlmostEqual( cache.read( 2.25, "obj2", "d" ).value, 2.4531, 3 )
		self.assertEqual( cache.read( 5, "obj2", "d" ), DoubleData( 16 ) )
		cache.setInterpolation( InterpolatedCache.Interpolation.Linear )
		self.assertEqual( cache.read( 5, "obj2", "d" ), DoubleData( 16 ) )
		cache.setInterpolation( InterpolatedCache.Interpolation.None )
		self.assertEqual( cache.read( 5, "obj2", "d" ), DoubleData( 16 ) )

	def testOldTransformationMatrixData( self ):

		cache = InterpolatedCache( "test/IECore/data/attributeCaches/transform.old.####.fio", interpolation = InterpolatedCache.Interpolation.Linear )
		self.assertEqual( cache.read( 4, ".parent", "transformCache.transform" ).value.rotate, Eulerd() )
		self.assertAlmostEqual( (cache.read( 4, ".pSphere1", "transformCache.transform" ).value.rotate - Eulerd( 0.244525, 0, 0 )).length(), 0, 2 )
		self.assertEqual( cache.read( 4.5, ".parent", "transformCache.transform" ).value.rotate, Eulerd() )
		self.assertAlmostEqual( (cache.read( 4.5, ".pSphere1", "transformCache.transform" ).value.rotate - Eulerd(0.283422, 0, 0)).length(), 0, 2 )

	def testThreading( self ) :
	
		def check( cache, frame ) :
		
			objects = cache.objects( frame )
			self.assertEqual( len( objects ), 2 )
			self.failUnless( "obj1" in objects )
			self.failUnless( "obj2" in objects )
			
			self.assertEqual( cache.attributes( frame, "obj1" ), [ "v3fVec" ] )
			
			attributes = cache.attributes( frame, "obj2" )
			self.assertEqual( len( attributes ), 2 )
			self.failUnless( "i" in attributes )
			self.failUnless( "d" in attributes )
			
			v = cache.read( frame, "obj1", "v3fVec" )
			self.failUnless( v[0].equalWithAbsError( V3f( frame ), 0.001 ) )
			self.failUnless( v[1].equalWithAbsError( V3f( frame ), 0.001 ) )
			self.failUnless( v[2].equalWithAbsError( V3f( frame ), 0.001 ) )
			self.assertEqual( cache.read( frame, "obj2", "i" ), IntData( int( frame ) ) )
			self.assertAlmostEqual( cache.read( frame, "obj2", "d" ).value, frame, 6 )
	
			cache.read( frame, "obj1" )
	
			self.failUnless( "testCache" in cache.headers( frame ) )
			
			h = cache.readHeader( frame, "testCache" )
			self.assertEqual( cache.readHeader( frame )["testCache"], h )
			
			self.failUnless( cache.contains( frame, "obj1" ) )
			self.failUnless( cache.contains( frame, "obj2" ) )
			self.failIf( cache.contains( frame, "ofsdfsdbj2" ) )

		self.__createCache()
		
		cache = InterpolatedCache(self.pathTemplate, interpolation = InterpolatedCache.Interpolation.Linear )
					
		threads = []
		for i in range( 0, 1000 ) :
			t = threading.Thread( target=check, args = ( cache, random.uniform( 0, 2 ) ) )
			t.start()
			threads.append( t )

		for t in threads :
			t.join()
	
	def testDefaultConstructor( self ) :
	
		self.__createCache()

		# If we provide a default constructor then it really oughn't to throw an exception
		cache = InterpolatedCache()
		# But if we try to do something with it before setting the template then it should complain
		self.assertRaises( RuntimeError, cache.readHeader, 0 )
		# And it should work as usual if we subsequently set the template
		cache.setPathTemplate( self.pathTemplate )
		cache.readHeader( 0 )
		self.assertEqual( cache.read( 0, "obj2", "d" ), DoubleData( 0 ) )

	def testReadMissing( self ) :

		# check that trying to read missing caches fails
		
		cache = InterpolatedCache( "iDontExist.######.fio" )
		self.assertRaises( RuntimeError, cache.read, 1.5, "a", "b" )
	
		# check that trying to read missing objects or attributes fails
	
		self.__createCache()
	
		cache = InterpolatedCache( self.pathTemplate )
		cache.read( 1.5, "obj2", "d" )
		
		self.assertRaises( RuntimeError, cache.read, 1.5, "iDontExist", "a" )
		self.assertRaises( RuntimeError, cache.read, 1.5, "obj2", "iDontExist" )

	def testMaxOpenFiles( self ) :

		cache = InterpolatedCache( self.pathTemplate, maxOpenFiles=20 )
		self.assertEqual( cache.getMaxOpenFiles(), 20 )
		
		cache.setMaxOpenFiles( 10 )
		self.assertEqual( cache.getMaxOpenFiles(), 10 )
		
	def tearDown(self):

		# cleanup
		allFiles = glob.glob( self.pathTemplate.replace( "####", "*" ) )
		for f in allFiles:
			os.remove( f )

if __name__ == "__main__":
	unittest.main()

